Add menu item to make the Studio to float on top of other windows

sejas opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

- Closes https://github.com/Automattic/dotcom-forge/issues/8985

## Proposed Changes

- Added a new menu item to make the Studio App floating over the rest of windows, to make easier to interact with the assistant and the window site.
- This option is not saved as a user preference and it's set as the default false (non floating) after Studio restarts, this respects how QuickTime works.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run `npm start`
- Click on View in the the top menu
- Observe the new menu item "Float on Top of All Other Windows"
- Click on the menu item
- Observe the Studio window floats over all other screens
- Click again in the menu item "Float on Top of All Other Windows" to deselect it
- Observe the Studio window does not float, and other windows can be on top of it

https://github.com/user-attachments/assets/a502b71b-ae8a-43e9-9367-56ad8d6b3337



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

sejas committed:
Add a new menu item to make the window float on top

fluiddot commented on src/menu.ts:
```diff
@@ -150,6 +150,16 @@ function getAppMenu( mainWindow: BrowserWindow | null ) {
 				{ role: 'togglefullscreen' },
+				{ type: 'separator' },
+				{
+					label: __( 'Float on Top of All Other Windows' ),
```
I wonder if this item should be under the Window tab instead of View. In fact, I noticed that in the designs it's under that section. WDYT?

<img width="339" alt="Screenshot 2024-09-05 at 12 55 50" src="https://github.com/user-attachments/assets/2197233b-7c1d-43ec-b657-7f34b84f6dd6">

katinthehatsite replied:
I had the same thought when testing this +1

sejas replied:
Great Observation,

I've decided to put the menu under View to follow QuickTime settings:

<img width="515" alt="quicktime" src="https://github.com/user-attachments/assets/9f3c327c-8b24-475a-a650-a3bd49af6d32">


I've checked other apps like VLC which displays that menu item under video instead of Window.

<img width="739" alt="vlc" src="https://github.com/user-attachments/assets/e1487a0e-2e1c-47c3-8c37-00299c22a105">

I've also observed that we don't have a "Window" menu item for Windows, we could add it but it would be the only item there.

![windows-menu](https://github.com/user-attachments/assets/18ef61b9-cd35-4959-9eb3-73749901fde7)

fluiddot replied:
Good point. To be honest, I don't have a strong preference between the View and Window tabs to place the item so looks good to me to put it under View. My comment was more regarding the discrepancy with the design.

fluiddot reviewed (approved):
LGTM 🎊 ! Tested on macOS and Windows.

I added a comment regarding where to place the new item but shouldn't block the PR.

sejas committed:
Fix toggle window floating for Windows

katinthehatsite reviewed (approved):
The changes look good and the functionality works as expected. It would be nice to move this to the Window section as suggested by another comment

sejas commented:
@fluiddot , @katinthehatsite , Thanks for reviewing it. I've updated the code to fix a bug in Windows, I would appreciate a second review.

I decided to put it under View to follow other apps like QuickTime. Let me know what you think. The change is trivial.

While testing on Windows, I've noticed the toggle was not working. Once the user clicked on the menu item, it was not "checked" and the window was floating constantly even after clicking the menu item for second or third time.

Fixed here: https://github.com/Automattic/studio/pull/511/commits/99908d1944205f1a9c573aeab053ec37a3f4607a

fluiddot commented:
> While testing on Windows, I've noticed the toggle was not working. Once the user clicked on the menu item, it was not "checked" and the window was floating constantly even after clicking the menu item for second or third time.

@sejas Ah, interesting. I didn't encounter this when I tested it on Windows. What Windows version have you tested it?

sejas commented:
@fluiddot , I'm using Windows 11 Home.

I've observed that the menu in Windows is being "rendered" inside the App, and that's probably why the auto-toggle is not working out of the box.

I encountered a couple of related issues:
- https://github.com/electron/electron/issues/10845
- https://github.com/electron/electron/issues/5055

They solve it by calling `trayIcon.setContextMenu `, but we are not using any [`Tray`](https://github.com/electron/electron/blob/4b5a83197911ce9de80004437d7ff41028e33bcd/docs/api/tray.md) which we don't. My understanding is that the menu is generated each time a user opens it in Windows.

fluiddot reviewed (approved):
LGTM 🎊 ! I've re-tested it in both macOS and Windows and worked nice.
